src/DescriptorsCommand.cc

```diff
@@ -409,94 +412,240 @@ void AddDescriptor::retrieve_aws_descriptorSet(const std::string &set_path) {
+
+    Json::Value constraints;
+    Json::Value name_arr;
+    name_arr.append("==");
+    name_arr.append(set_name);
+    constraints[VDMS_DESC_SET_NAME_PROP] = name_arr;
+
+    bool unique = true;
+
+    // Query set node
+    query.QueryNode(set_ref, VDMS_DESC_SET_TAG, link, constraints, results,
+                    unique);
+
+    if (cmd.isMember("link")) {
+        add_link(query, cmd["link"], node_ref, VDMS_DESC_EDGE_TAG);
+    }
+
+    Json::Value props_edge;
+    query.AddEdge(-1, set_ref, node_ref, VDMS_DESC_SET_EDGE_TAG, props_edge);
+
+    // TODO: deleting files here causes problems with concurrency (TestRetail.py)
+    // keeping local copies as a temporary solution
+    // if(_use_aws_storage)
+    // {
+    //     //delete files in set_path
+    //     std::uintmax_t n = fs::remove_all(set_path);
+    //     std::cout << "Deleted " << n << " files or directories\n";
+    // }
+
+    return 0;
 
+}
+
+int AddDescriptor::add_descriptor_batch(PMGDQuery &query,
+                                         const Json::Value &jsoncmd,
+                                         const std::string &blob, int grp_id,
+                                         Json::Value &error){
+
+    int expected_blb_size;
+    int nr_expected_descs;
+    int dimensions;
+
+    //Extract set name
+    const Json::Value &cmd = jsoncmd[_cmd_name];
+    const std::string set_name = cmd["set"].asString();
+
+    Json::Value props = get_value<Json::Value>(cmd, "properties");
+
+    //extract properties list and get filepath/object location of set
+    Json::Value prop_list = get_value<Json::Value>(cmd, "batch_properties");
+    const std::string set_path = get_set_path(query, set_name, dimensions);
+
+    if (set_path.empty()) {
+        error["info"] = "Set " + set_name + " not found";
+        error["status"] = RSCommand::Error;
+        return -1;
+    }
+
+    std::string label = get_value<std::string>(cmd, "label", "None");
+    props[VDMS_DESC_LABEL_PROP] = label;
+
+    // retrieve the descriptor set from AWS here
+    // operations are currently done in memory with no subsequent write to disk
+    // so there's no need to re-upload to AWS
     if (_use_aws_storage) {
-      // delete files in set_path
-      std::uintmax_t n = fs::remove_all(set_path);
-      std::cout << "Deleted " << n << " files or directories\n";
+        retrieve_aws_descriptorSet(set_path);
     }
 
-    return -1;
-  }
+    // Note dimensionse are based on a 32 bit integer, hence the /4 math on size
+    // as the string blob is sized in 8 bit ints.
+    nr_expected_descs = prop_list.size();
+    expected_blb_size = nr_expected_descs * dimensions * 4;
+
+    //Verify length of input is matching expectations
+    if (blob.length() != expected_blb_size) {
+        std::cerr << "AddDescriptor::insert_descriptor: ";
+        std::cerr << "Expectected Blob Length Does Not Match Input ";
+        std::cerr << blob.length() << " != " << expected_blb_size << std::endl;
```

Comment:
The 2 lengths will show as numbers in the output & might not be clear which is expected & which is the actual length. Might be better to add a tag to each output number to mark this.